File: worker/src/protocol.ts
// Wire protocol between game clients and a GameRoom Durable Object.
// One JSON object per WebSocket text frame. Field names are kept short
// because position updates dominate traffic.
//
// The C# twin of this file is EightPlayers/EcsNet/Protocol.cs — keep them in sync.

export const PROTO_VERSION = 1

export type Components = Record<string, unknown>

// ---- client -> server ----

export interface HelloMsg {
  t: 'hello'
  proto: number
  name: string
}

export interface SpawnReq {
  t: 'spawn'
  /** client-chosen temporary id, echoed back so the client can map it to the server entity id */
  tmp: number
  components: Components
}

export interface SetReq {
  t: 'set'
  e: number
  components: Components
}

export interface DespawnReq {
  t: 'despawn'
  e: number
}

export interface PingMsg {
  t: 'ping'
  ts: number
}

/**
 * Update the room's shared world. seed: first write wins.
 * num: the party's current level, monotonically increasing.
 */
export interface WorldReq {
  t: 'world'
  seed?: string
  num?: number
}

/**
 * Fire-and-forget world event (door opened, effect triggered...). Relayed to
 * every other client, never stored — late joiners reconstruct state from the
 * deterministic world, not the event stream.
 */
export interface EventReq {
  t: 'event'
  kind: string
  data?: unknown
}

export type ClientMsg = HelloMsg | SpawnReq | SetReq | DespawnReq | PingMsg | WorldReq | EventReq

// ---- server -> client ----

export interface EntityRecord {
  e: number
  owner: number
  components: Components
}

export interface WelcomeMsg {
  t: 'welcome'
  proto: number
  you: number
  room: string
  peers: { id: number; name: string }[]
  snapshot: EntityRecord[]
  /** null until some client claims the room's world seed */
  world: { seed: string; num: number } | null
}

export interface WorldEvt {
  t: 'world'
  seed: string
  num: number
}

export interface SpawnEvt {
  t: 'spawn'
  e: number
  owner: number
  components: Components
  /** present only on the copy sent to the owner */
  tmp?: number
}

export interface SetEvt {
  t: 'set'
  e: number
  components: Components
}

export interface DespawnEvt {
  t: 'despawn'
  e: number
}

export interface PeerEvt {
  t: 'peer'
  id: number
  name: string
  joined: boolean
}

export interface PongMsg {
  t: 'pong'
  ts: number
}

export interface ErrorMsg {
  t: 'error'
  message: string
}

export interface EventEvt {
  t: 'event'
  from: number
  kind: string
  data?: unknown
}

export type ServerMsg =
  | WelcomeMsg
  | SpawnEvt
  | SetEvt
  | DespawnEvt
  | PeerEvt
  | PongMsg
  | ErrorMsg
  | WorldEvt
  | EventEvt

File: EightPlayers/EcsNet/Protocol.cs
using Newtonsoft.Json.Linq;

namespace EightPlayers.EcsNet
{
    // C# twin of worker/src/protocol.ts — keep the two in sync.
    // Outgoing messages are built as JObjects; incoming frames are parsed into
    // a single ServerMsg record with only the fields that message type uses.

    public static class Protocol
    {
        public const int Version = 1;

        public static string Hello(string name) =>
            new JObject { ["t"] = "hello", ["proto"] = Version, ["name"] = name }.ToString(Newtonsoft.Json.Formatting.None);

        public static string Spawn(int tmp, JObject components) =>
            new JObject { ["t"] = "spawn", ["tmp"] = tmp, ["components"] = components }.ToString(Newtonsoft.Json.Formatting.None);

        public static string Set(int entity, JObject components) =>
            new JObject { ["t"] = "set", ["e"] = entity, ["components"] = components }.ToString(Newtonsoft.Json.Formatting.None);

        public static string Despawn(int entity) =>
            new JObject { ["t"] = "despawn", ["e"] = entity }.ToString(Newtonsoft.Json.Formatting.None);

        public static string Ping(long ts) =>
            new JObject { ["t"] = "ping", ["ts"] = ts }.ToString(Newtonsoft.Json.Formatting.None);

        public static string World(string seed) =>
            new JObject { ["t"] = "world", ["seed"] = seed }.ToString(Newtonsoft.Json.Formatting.None);

        public static string WorldNum(int num) =>
            new JObject { ["t"] = "world", ["num"] = num }.ToString(Newtonsoft.Json.Formatting.None);

        public static string Event(string kind, JObject data) =>
            new JObject { ["t"] = "event", ["kind"] = kind, ["data"] = data }.ToString(Newtonsoft.Json.Formatting.None);

        public static JObject PosComponent(float x, float y) =>
            new JObject { ["pos"] = new JObject { ["x"] = x, ["y"] = y } };

        public static JObject HpComponent(float cur, float max) =>
            new JObject { ["hp"] = new JObject { ["cur"] = cur, ["max"] = max } };

        public static JObject LevelComponent(int seed, int num) =>
            new JObject { ["level"] = new JObject { ["seed"] = seed, ["num"] = num } };

        public static JObject NpcComponents(int index, string type, float x, float y) =>
            new JObject
            {
                ["npc"] = new JObject { ["i"] = index, ["type"] = type },
                ["pos"] = new JObject { ["x"] = x, ["y"] = y },
            };

        public static JObject PlayerComponents(string name, int color, string charType, float x, float y, float hp, float hpMax) =>
            new JObject
            {
                ["player"] = new JObject { ["name"] = name, ["color"] = color, ["char"] = charType },
                ["pos"] = new JObject { ["x"] = x, ["y"] = y },
                ["hp"] = new JObject { ["cur"] = hp, ["max"] = hpMax },
            };
    }

    public sealed class ServerMsg
    {
        public string T;
        public int You;           // welcome
        public string Room;       // welcome
        public JArray Snapshot;   // welcome
        public JArray Peers;      // welcome
        public int Entity;        // spawn/set/despawn
        public int Owner;         // spawn
        public int Tmp = -1;      // spawn (only on the owner's copy)
        public JObject Components; // spawn/set
        public int PeerId;        // peer
        public string PeerName;   // peer
        public bool Joined;       // peer
        public long Ts;           // pong
        public string Message;    // error
        public string WorldSeed;  // welcome (null if unset) / world
        public int WorldLevel = 1; // welcome / world
        public string Kind;       // event
        public JObject EventData; // event
        public int From;          // event

        public static ServerMsg Parse(string raw)
        {
            var jo = JObject.Parse(raw);
            var msg = new ServerMsg { T = (string)jo["t"] };
            switch (msg.T)
            {
                case "welcome":
                    msg.You = (int)jo["you"];
                    msg.Room = (string)jo["room"];
                    msg.Snapshot = (JArray)jo["snapshot"];
                    msg.Peers = (JArray)jo["peers"];
                    if (jo["world"] is JObject world)
                    {
                        msg.WorldSeed = (string)world["seed"];
                        msg.WorldLevel = (int?)world["num"] ?? 1;
                    }
                    break;
                case "world":
                    msg.WorldSeed = (string)jo["seed"];
                    msg.WorldLevel = (int?)jo["num"] ?? 1;
                    break;
                case "event":
                    msg.Kind = (string)jo["kind"];
                    msg.EventData = jo["data"] as JObject;
                    msg.From = (int?)jo["from"] ?? 0;
                    break;
                case "spawn":
                    msg.Entity = (int)jo["e"];
                    msg.Owner = (int)jo["owner"];
                    msg.Components = (JObject)jo["components"];
                    if (jo["tmp"] != null) msg.Tmp = (int)jo["tmp"];
                    break;
                case "set":
                    msg.Entity = (int)jo["e"];
                    msg.Components = (JObject)jo["components"];
                    break;
                case "despawn":
                    msg.Entity = (int)jo["e"];
                    break;
                case "peer":
                    msg.PeerId = (int)jo["id"];
                    msg.PeerName = (string)jo["name"];
                    msg.Joined = (bool)jo["joined"];
                    break;
                case "pong":
                    msg.Ts = (long)jo["ts"];
                    break;
                case "error":
                    msg.Message = (string)jo["message"];
                    break;
            }
            return msg;
        }
    }
}
